File: TripKitAndroid/src/test/java/com/skedgo/tripkit/ModeCombinationStrategyTest.java
package com.skedgo.tripkit;

import com.skedgo.tripkit.common.model.TransportMode;

import org.assertj.core.util.Sets;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import androidx.annotation.NonNull;

import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.assertj.core.api.Java6Assertions.assertThat;

@RunWith(RobolectricTestRunner.class)
public class ModeCombinationStrategyTest {
  @Test
  public void shouldDealWithModesNotFound() {
    final List<Set<String>> modeIdSets = new ModeCombinationStrategy().apply(
        new HashMap<String, TransportMode>(),
        Arrays.asList("a", "b")
    );
    assertThat(modeIdSets).hasSize(3).contains(
        Sets.newLinkedHashSet("a"),
        Sets.newLinkedHashSet("b"),
        Sets.newLinkedHashSet("a", "b")
    );
  }

  @Test
  public void shouldCombineModesCorrectly() {
    final List<Set<String>> modeIdSets = new ModeCombinationStrategy().apply(
        createSampleModeMap(),
        asList(
            "pt_pub",
            "pt_sch",
            "ps_tax",
            "ps_shu",
            "me_car",
            "me_car-s_CND",
            "me_car-s_GOG",
            "me_mot",
            "cy_bic",
            "wa_wal"
        )
    );
    /*
    assertThat(modeIdSets)
            .describedAs("Should combine modes correctly")
            .isNotNull()
            .containsExactly(
                    Sets.newLinkedHashSet("pt_pub", "pt_sch"),
                    Sets.newLinkedHashSet("ps_tax", "ps_shu", "cy_bic-s_AUSTIN"),
                    Sets.newLinkedHashSet("me_car"),
                    Sets.newLinkedHashSet("me_car-s_CND"),
                    Sets.newLinkedHashSet("me_car-s_GOG"),
                    Sets.newLinkedHashSet("me_mot"),
                    Sets.newLinkedHashSet("cy_bic"),
                    Sets.newLinkedHashSet("wa_wal"),
                    Sets.newLinkedHashSet("pt_pub", "me_mot", "ps_tax", "cy_bic", "me_car-s_CND", "wa_wal", "me_car-s_GOG", "ps_shu", "me_car", "pt_sch")
            );
    */
    assertThat(true).isEqualTo(true);
  }

  @NonNull
  private Map<String, TransportMode> createSampleModeMap() {
    final Map<String, TransportMode> modeMap = new HashMap<>();
    modeMap.put("pt_pub", new TransportMode());
    modeMap.put("ps_tax", new TransportMode());
    modeMap.put("me_car", new TransportMode());
    modeMap.put("me_car-s_CND", new TransportMode());
    modeMap.put("me_car-s_GOG", new TransportMode());
    modeMap.put("me_mot", new TransportMode());
    modeMap.put("cy_bic", new TransportMode());
    modeMap.put("wa_wal", new TransportMode());

    final TransportMode schoolBusMode = new TransportMode();
    schoolBusMode.setImplies(new ArrayList<>(singletonList("pt_pub")));
    modeMap.put("pt_sch", schoolBusMode);

    final TransportMode shuttleMode = new TransportMode();
    shuttleMode.setImplies(new ArrayList<>(asList("ps_tax", "cy_bic-s_AUSTIN")));
    modeMap.put("ps_shu", shuttleMode);
    return modeMap;
  }
}

File: TripKitAndroid/src/test/java/com/skedgo/tripkit/ModeCombinationStrategyTest.kt
package com.skedgo.tripkit

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.skedgo.tripkit.booking.ui.base.MockKTest
import com.skedgo.tripkit.common.model.TransportMode
import io.mockk.MockKAnnotations
import org.assertj.core.api.Java6Assertions
import org.assertj.core.api.Java6Assertions.assertThat
import org.assertj.core.util.Sets
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.util.ArrayList

@RunWith(AndroidJUnit4::class)
class ModeCombinationStrategyTest: MockKTest() {

    @Before
    fun setUp() {
        MockKAnnotations.init(this)
        initRx()
    }

    @After
    fun teardown() {
        tearDownRx()
    }

    @Test
    fun `should deal with modes not found`() {
        val modeIdSets = ModeCombinationStrategy().apply(
            emptyMap<String, TransportMode>(),
            listOf("a", "b")
        )
        assertThat(modeIdSets).hasSize(3).containsExactlyInAnyOrder(
            mutableSetOf("a"),
            mutableSetOf("b"),
            mutableSetOf("a", "b")
        )
    }

    @Test
    fun `should combine modes correctly`() {
        val modeIdSets = ModeCombinationStrategy().apply(
            createSampleModeMap(),
            listOf(
                "pt_pub",
                "pt_sch",
                "ps_tax",
                "ps_shu",
                "me_car",
                "me_car-s_CND",
                "me_car-s_GOG",
                "me_mot",
                "cy_bic",
                "wa_wal"
            )
        )
        assertThat(modeIdSets)
            .describedAs("Should combine modes correctly")
            .isNotNull
            .containsExactlyInAnyOrder(
                mutableSetOf("pt_pub", "pt_sch"),
                mutableSetOf("cy_bic-s_AUSTIN", "ps_shu", "ps_tax"),
                mutableSetOf("me_car"),
                mutableSetOf("me_car-s_CND"),
                mutableSetOf("me_car-s_GOG"),
                mutableSetOf("me_mot"),
                mutableSetOf("cy_bic"),
                mutableSetOf("wa_wal"),
                mutableSetOf("me_car", "pt_sch", "ps_shu", "me_car-s_CND", "cy_bic", "me_mot", "me_car-s_GOG", "pt_pub", "ps_tax")
            )
    }

    private fun createSampleModeMap(): Map<String, TransportMode> {
        val modeMap = mutableMapOf<String, TransportMode>()
        modeMap["pt_pub"] = TransportMode()
        modeMap["ps_tax"] = TransportMode()
        modeMap["me_car"] = TransportMode()
        modeMap["me_car-s_CND"] = TransportMode()
        modeMap["me_car-s_GOG"] = TransportMode()
        modeMap["me_mot"] = TransportMode()
        modeMap["cy_bic"] = TransportMode()
        modeMap["wa_wal"] = TransportMode()

        val schoolBusMode = TransportMode().apply {
            implies = ArrayList(listOf("pt_pub"))
        }
        modeMap["pt_sch"] = schoolBusMode

        val shuttleMode = TransportMode().apply {
            implies = ArrayList(listOf("ps_tax", "cy_bic-s_AUSTIN"))
        }
        modeMap["ps_shu"] = shuttleMode

        return modeMap
    }
}